feat(ngMessages): add support for default message

gdangelo opened the pull request:
Added support for showing default message when no values are mapped with ng-message.

Closes #12008

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/angular/angular.js/12213)

<!-- Reviewable:end -->

gdangelo committed:
feat(ngMessages): add support for default message

Added support for showing default message when no values are mapped with ng-message.

Closes #12008

lgalfaso commented:
it looks like this PR adds quite a few of duplicated code. Before doing a proper review, would it be possible to try to keep the core a little bit more DRY?

gdangelo commented:
I will make the modifications soon.

gdangelo committed:
refactor(ngMessages): change code to make it DRY

Added a new param (boolean, isDefault) to the factory ngMessageDirectiveFactory, in order to reuse code in ngMessageDefault directive.

Closes #12008

petebacondarwin commented:
This looks reasonable let's get it into 1.6 early as soon as 1.5.0 is released.

mikeycgto commented:
Any updates on this making it into 1.6?

gkalpak commented on src/ngMessages/messages.js:
```diff
@@ -239,6 +239,19 @@ var jqLite = angular.element;
+ * <div ng-messages="myForm.myField.$error" role="alert">
+ *   <div ng-message="required">This field is required</div>
+ *   <div ng-message="minlength">This field is too short</div>
+ *   <div ng-message-default>This is a default message</div>
```
I would use a more meaningful message here to give a better impression of how to use the directive.

mikeycgto replied:
Anything in specific that should be mentioned here?

gkalpak replied:
Anything that you would use in an actual app. In this case, since these seem to be validation messages, the default message would be something like 'Valid' or a checkmark or 'This field is valid` etc.

gkalpak commented on src/ngMessages/messages.js:
```diff
@@ -379,9 +398,17 @@ angular.module('ngMessages', [])
-              ? $animate.setClass($element, ACTIVE_CLASS, INACTIVE_CLASS)
-              : $animate.setClass($element, INACTIVE_CLASS, ACTIVE_CLASS);
+           if (unmatchedMessages.length !== totalMessages) {
+             // Unset default message if setted
```
_setted_  -->  _set_

mikeycgto replied:
Not sure I follow...?

gkalpak replied:
`X  -->  Y` is my way of telling that there is a typo in `X`, it should be changed to `Y` :stuck_out_tongue: 
(Basically, change `setted` to `set`.)

gkalpak commented on src/ngMessages/messages.js:
```diff
@@ -397,23 +424,32 @@ angular.module('ngMessages', [])
-           latestKey++;
+         this.register = function(comment, messageCtrl, isDefault) {
+           if (isDefault) {
+            ctrl.default = messageCtrl;
```
Wrong indentation (here and below).

gkalpak commented on src/ngMessages/messages.js:
```diff
@@ -397,23 +424,32 @@ angular.module('ngMessages', [])
-           delete messages[key];
+         this.deregister = function(comment, isDefault) {
+           if (isDefault) {
+            delete ctrl.default;
```
I think you also need to detach `ctrl.default` (if defined), because if it is visible, it won't get removed by `ctrl.reRender()`.

mikeycgto replied:
Do you mean to call `removeMessageNode` here for the default?

gkalpak replied:
I don't think `removeMessageNode()` works for `ctrl.default`. I was thinking `ctrl.default.detach()`.
<sub>And there should also be a test about dynamically adding/removing an `<ng-message-default>` element (e.g. using `ngIf`).</sub>

gkalpak commented on src/ngMessages/messages.js:
````diff
@@ -594,35 +630,74 @@ angular.module('ngMessages', [])
+    * </ng-messages>
+    * ```
+    *
+    * @param {expression} ngMessageDefault|when no ngMessage matches.
````
It doesn't take any params.

gkalpak commented on test/ngMessages/messagesSpec.js:
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
+                       '</div>')($rootScope);
+    $rootScope.$digest();
+
+    //expect(element.text()).not.toContain('Message is set');
```
Remove comment.

gkalpak commented on test/ngMessages/messagesSpec.js:
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
+    $rootScope.$digest();
+
+    //expect(element.text()).not.toContain('Message is set');
+    expect(element.text()).toContain('Default message is set');
```
Why not use `.toBe()`?

gkalpak commented on test/ngMessages/messagesSpec.js:
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
+
+    $rootScope.$apply(function() {
+      $rootScope.col = { val: true };
+    });
```
`$rootScope.$apply('col = {val: true}')` is more concise :smiley:

gkalpak commented on test/ngMessages/messagesSpec.js:
```diff
@@ -401,6 +401,24 @@ describe('ngMessages', function() {
+    });
+
+    //expect(element.text()).toContain('Message is set');
+    expect(element.text()).not.toContain('Default message is set');
```
Again, why not use `.toBe()` and verify the exact contents?

gkalpak commented:
A few more tests would be (e.g. using `ng-messages-multiple` or using `ngIf` on `ngMessageDefault` etc). But overall looks good.

gdangelo commented:
Hi @gkalpak, @mikeycgto, thanks to follow this feature request. Do we need anything else to include the feature in the 1.6 release?

petebacondarwin commented:
@gdangelo - yes, @gkalpak provided a number of comments that need addressing, including some more tests.
